src/Catalog/Dnx/DnxMaker.cs

```diff
@@ -260,6 +274,122 @@ private async Task<Uri> SaveNupkgAsync(Stream nupkgStream, Storage storage, stri
             return destinationUri;
         }
 
+        private async Task CopyIconFromAzureStorageIfExistAsync(
+            IAzureStorage sourceStorage,
+            Storage destinationStorage,
+            string packageId,
+            string normalizedPackageVersion,
+            string iconFilename,
+            CancellationToken cancellationToken)
+        {
+            await CopyIconAsync(
+                iconFilename,
+                destinationStorage,
+                packageId,
+                normalizedPackageVersion,
+                cancellationToken,
+                () => GetPackageStreamAsync(sourceStorage, packageId, normalizedPackageVersion, cancellationToken),
+                disposeOfPackageStream: true);
+        }
+
+        private async Task CopyIconFromNupkgStreamAsync(
+            Stream nupkgStream,
+            string iconFilename,
+            Storage destinationStorage,
+            string packageId,
+            string normalizedPackageVersion,
+            CancellationToken cancellationToken)
+        {
+            await CopyIconAsync(
+                iconFilename,
+                destinationStorage,
+                packageId,
+                normalizedPackageVersion,
+                cancellationToken,
+                () => Task.FromResult(nupkgStream),
+                disposeOfPackageStream: false);
+        }
+
+        private async Task CopyIconAsync(
```

Comment:
>CopyIconAsync [](start = 27, length = 13)

I feel as if this method would be cleaner if the package stream was managed by its caller. The caller can create the stream and choose whether or not to dispose it.